tests/filesystem/ProtectedFileControllerTest.php

```diff
@@ -0,0 +1,201 @@
+<?php
+
+use SilverStripe\Filesystem\Storage\AssetStore;
+use SilverStripe\Filesystem\Storage\ProtectedFileController;
+
+class ProtectedFileControllerTest extends FunctionalTest {
+
+	protected static $fixture_file = 'FileTest.yml';
+
+	public function setUp() {
+		parent::setUp();
+
+		// Set backend root to /ImageTest
+		AssetStoreTest_SpyStore::activate('ProtectedFileControllerTest');
+
+		// Create a test folders for each of the fixture references
+		foreach (Folder::get() as $folder) {
+			/** @var Folder $folder */
+			$filePath = AssetStoreTest_SpyStore::getLocalPath($folder);
+			\Filesystem::makeFolder($filePath);
+		}
+
+		// Create a test files for each of the fixture references
+		foreach (File::get()->exclude('ClassName', 'Folder') as $file) {
+			/** @var File $file */
+			$path = AssetStoreTest_SpyStore::getLocalPath($file);
+			\Filesystem::makeFolder(dirname($path));
+			$fh = fopen($path, "w+");
+			fwrite($fh, str_repeat('x', 1000000));
+			fclose($fh);
+
+			// Create variant for each file
+			$this->getAssetStore()->setFromString(
+				str_repeat('y', 100),
+				$file->Filename,
+				$file->Hash,
+				'variant'
+			);
+		}
+	}
+
+	/**
+	 * @dataProvider getFilenames
+	 */
+	public function testIsValidFilename($name, $isValid) {
+		$controller = new ProtectedFileController();
+		$this->assertEquals(
+			$isValid,
+			$controller->isValidFilename($name),
+			"Assert filename \"$name\" is " . $isValid ? "valid" : "invalid"
+		);
+	}
+
+	public function getFilenames() {
+		return array(
+			// Valid names
+			array('name.jpg', true),
+			array('parent/name.jpg', true),
+			array('parent/name', true),
+			array('parent\name.jpg', true),
+			array('parent\name', true),
+			array('name', true),
+
+			// Invalid names
+			array('.invalid/name.jpg', false),
+			array('.invalid\name.jpg', false),
+			array('.htaccess', false),
+			array('test/.htaccess.jpg', false),
+			array('name/.jpg', false),
+			array('test\.htaccess.jpg', false),
+			array('name\.jpg', false)
+		);
+	}
+
+	/**
+	 * Test that certain requests are denied
+	 */
+	public function testRequestDenied() {
+		$result = $this->get('assets/.protected/file.jpg');
+		$this->assertResponseEquals(400, null, $result);
```

Comment:
I think you are looking at it the wrong way; It returns a 400 if the REQUEST is invalid. That is, it contains a `.name` directory.

It doesn't check if the file exists, since it is refusing to process an invalid request. :)